### PR TITLE
charmstore: add supported series field to entity

### PR DESCRIPTION
**Pull request by rogpeppe**
We also fix the bug where we couldn't start the charm store because
we'd deleted all the old migrations.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1178/
Test PASSed.

**Comment by mhilton** [comment]
LGTM

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1179/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1180/
Test PASSed.

File: internal/charmstore/migrations.go [hunk 1]
```diff
@@ -20,7 +22,24 @@ import (
 // migration name and function to this list, and update the
 // TestMigrateMigrationList test in migration_test.go adding the new name(s).
 // Note that migration names must be unique across the list.
-var migrations = []migration{}
+//
+// A migration entry may have a nil migration function if the migration
+// is obsolete. Obsolete migrations should never be removed entirely,
+// otherwise the charmstore will see the old migrations in the table
+// and refuse to start up because it thinks that it's running an old
+// version of the charm store on a newer version of the database.
```
Comment (jrwren) [inline internal/charmstore/migrations.go:30]: Excellent explanation. I wasn't sure after standup. This is very clear. TY

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -566,6 +561,33 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 	return nil
 }
 
+// denormalizeEntity sets all denormalized fields in e
+// from their associated canonical fields.
+//
+// It is the responsibility of the caller to set e.SupportedSeries
+// if the entity URL does not contain a series.
```
Comment (jrwren) [inline internal/charmstore/store.go:568]: Maybe document that e.SupportedSeries is overwritten with e.URL.Series even if e.SupportSeries is not default?

**Comment by jrwren** [comment]
:+1: I like that test setup structs are shorter now too. It reads a bit easier. Thanks.

File: internal/charmstore/store.go (hunk 2 above)
Comment (rogpeppe) [inline internal/charmstore/store.go:568]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1181/
Test FAILed.

**Pull request merged**